### PR TITLE
gcp - appengine initial setup with nested resources

### PR DESCRIPTION
**Pull request by alexkarpitski**
Add `query` and `get` methods for `App Engine` resources.

### CODE REVIEW

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+# limitations under the License.
+
+import re
+
+from c7n_gcp.provider import resources
+from c7n_gcp.query import QueryResourceManager, TypeInfo, ChildResourceManager, ChildTypeInfo
+from c7n.utils import local_session
+
+
+@resources.register('appengine-app')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: ditch the -app suffix, also consistent hyphenation app-engine

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
+# Copyright 2019 Capital One Services, LLC
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: Unless your doing work under contract / employment paid for by capitalone, this isn't correct. This project uses contributor license agreements from apache foundation, which do not use copyright assignment. Hence these files should reflect the entity/organization of the author to be accurate.

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        version = 'v1'
+        component = 'apps'
+        enum_spec = ('get', '[@]', None)
+        scope = None
+        id = 'id'
+
+        @staticmethod
+        def get(client, resource_info):
+            return client.execute_query(
+                'get', {'appsId': re.compile('apps/(.*)').match(
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: there is no point to compiling a regex if your throwing away the compilation. either store the compiled result as a class var, or ditch the compilation.

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    def get_resource_query(self):
+        return {'appsId': local_session(self.session_factory).get_default_project()}
+
+
+@resources.register('appengine-certificate')
+class AppEngineCertificate(ChildResourceManager):
+
+    def _get_parent_resource_info(self, child_instance):
+        return {'resourceName': re.compile(
+            '(apps/.*?)/authorizedCertificates/.*').match(child_instance['name']).group(1)}
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: there is no point to compiling a regex if your throwing away the compilation. either store the compiled result as a class var, or ditch the compilation.

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+        parent_spec = {
+            'resource': 'appengine-app',
+            'child_enum_params': {
+                ('id', 'appsId')
+            }
+        }
+
+        @staticmethod
+        def get(client, resource_info):
+            name_param_re = re.compile('apps/(.*?)/authorizedCertificates/(.*)')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: there is no point to compiling a regex if your throwing away the compilation. either store the compiled result as a class var, or ditch the compilation.

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+        parent_spec = {
+            'resource': 'appengine-app',
+            'child_enum_params': {
+                ('id', 'appsId')
+            }
+        }
+
+        @staticmethod
+        def get(client, resource_info):
+            rule_re = re.compile('apps/(.*?)/firewall/ingressRules/(.*)')
```
Comment (kapilt) [inline tools/c7n_gcp/c7n_gcp/resources/appengine.py]: there is no point to compiling a regex if your throwing away the compilation. either store the compiled result as a class var, or ditch the compilation.

File: tools/c7n_gcp/c7n_gcp/resources/appengine.py (hunk 2 above)
Comment (alexkarpitski) [reply tools/c7n_gcp/c7n_gcp/resources/appengine.py]: @jondavidi, @jonathana, @evhenyshaliov - could you advice whether it's right to keep it or not from the legal perspective?

**Comment by kapilt** [comment]
all of the resources here besides app-engine feel more appropriate as filters rather than resource types.

**Comment by kapilt** [comment]
wrt to some of these gcp contributions, it would be helpful to have use cases sent in as part of the pull request, a lot of this is reverse engineering intent and usability and applicability to custodian from the api docs. sending some example use cases that these are intended to suffice would simplify and hopefully clarify the utility or not.

**Comment by alexkarpitski** [comment]
> 
> 
> wrt to some of these gcp contributions, it would be helpful to have use cases sent in as part of the pull request, a lot of this is reverse engineering intent and usability and applicability to custodian from the api docs. sending some example use cases that these are intended to suffice would simplify and hopefully clarify the utility or not.

Sure, here are some cases the proposed resources could be used in:
- `app-engine` - is a parent resource for the rest ones (required)
- `app-engine-certificate` - check and notify users if a certificate is about to expire
- `app-engine-domain-mapping` - remove mappings that are no longer in use
- `app-engine-firewall-ingress-rule` - notify if a required rule is not set up

Regarding `app-engine-domain`, it contains information about the domain that is parent to `app-engine-domain-mapping`. However, due to not having any methods but [list](https://cloud.google.com/appengine/docs/admin-api/reference/rest/v1/apps.authorizedDomains/list) it could be skipped.

**Comment by kapilt** [comment]
the domain mapping no longer in use seems suspect as a use case, whats no longer in use is open to interpretation that a policy can't express directly except as a static value. however looking for particular domains does seem useful.

**Pull request merged**